src/Development/IDE/Plugin.hs

```diff
@@ -0,0 +1,33 @@
+
+module Development.IDE.Plugin(Plugin(..), codeActionPlugin) where
+
+import Data.Default
+import Development.Shake
+import Development.IDE.LSP.Server
+
+import           Language.Haskell.LSP.Types
+import Development.IDE.Core.Rules
+import qualified Language.Haskell.LSP.Core as LSP
+import Language.Haskell.LSP.Messages
+
+
+data Plugin = Plugin
+    {pluginRules :: Rules ()
+    ,pluginHandler :: PartialHandlers
+    }
+
+instance Default Plugin where
+    def = Plugin mempty def
+
+instance Semigroup Plugin where
+    Plugin x1 y1 <> Plugin x2 y2 = Plugin (x1<>x2) (y1<>y2)
+
+instance Monoid Plugin where
+    mempty = def
+
+
+codeActionPlugin :: (LSP.LspFuncs () -> IdeState -> TextDocumentIdentifier -> Range -> CodeActionContext -> IO [CAResult]) -> Plugin
```

Comment:
Why is `codeActionPlugin` in here instead of `Development.IDE.Plugin.CodeAction`?

Reply:
The idea is that `codeActionPlugin` is the general helper for writing code actions, so when there are two code action plugins, they both use this helper, but have different internals. It's closer to the model HIE follows. The aim would eventually for there to be 5 or 6 such helpers in this module.

Reply:
Makes nense.